nitrite/src/main/java/org/dizitart/no2/collection/NitriteDocument.java

```diff
@@ -71,6 +69,21 @@ public Document put(String field, Object value) {
             throw new ValidationException("Type " + value.getClass().getName()
                 + " does not implement java.io.Serializable");
         }
+    }
+
+    @Override
+    public Document put(String key, Object value, boolean ignoreFieldSeparator) {
```

Comment:
How are you planning to get the value if you are putting values ignoring fieldSeparator? Why can't you use a different field separator than the default one?

Reply:
Package Search is indexing a lot of data from repositories, including Maven Central. Nitrite is used as local caches for our plugin. As such, the strings in the data might contains any character, hence triggering the `deepPut` unwillingly.

Changing the shape of the document will mess up the deserialization using kotlinx.serialization. As such we need a way to serialize without [changing the shape of the document](https://github.com/nitrite/nitrite-java/pull/969/commits/f57a0daf719d01d7bf3beb2dcb1acbf0fd11c34d#diff-70a7e40d93f864fbf7ae03cdca2dedd8c8a577ba82b837d17c815a4ca4589ea8R154).

On a side note, I would guess that this `fieldSeparator` "optimization" has been designed for a better way to store data on disk. If so, I would instead recommend to have `NitriteDatabase` to handle the transformation from  a "non-nested document" to a "nested" one right before it gets stored and vice versa when readingfrom storage. This will ensure the field separator mechanism to be transparent from `Document` initialization and will allow the `NitriteConfig.fieldSeparator` to not be `static` (which to me looks kinda of a hack to let any document access it regardless of the configuration of the DB it will stored in). 

I would also stress this `fieldSeparator` transformation in the docs since it might change the shape of the queries one might do.

Alternatively, you might want to allow to disable such mechanism altogether just like for #966

Reply:
I am failing to understand the issue you are facing here. If for your data, `deepPut` is getting invoked unintentionally, that means, you have "." character in your field, which should not be the case as naming a field is under your complete control only. If the value contains any arbitrary string, it will not trigger `deepPut`. In other words, if your data can be converted to valid json, you should not be facing any issue in nitrite document also.

To better understand your issue, could you please provide me some example data, your kotlin entity class and the document it is generating. May be I'll be able to point out a better solution for you.

The`fieldSeparator` is not designed for better storage. It is there to query nested documents in a collection, similar to mongo. Using it to put or get nested document is a convenience, nothing more.

The reason for `NitriteConfig.fieldSeparator` being a static field is to make it a global configuration application wide. So that all collections, indexes have uniform convention to access data from a document. It minimizes the configuration burden and unintentional bugs in the code, if user choose to customize the `fieldSeparator`.

Reply:
Have a look [here](https://github.com/lamba92/nitrite-java/blob/a911586788d89865153eabc74e75bf5a13f35397/potassium-nitrite/src/test/kotlin/org/dizitart/kno2/KotlinXSerializationMapperTest.kt#L223-L249). As I mentioned before, it makes sense that at a storage level the document is normalized using the deep put. I see the point. But it is incorrect that the act itself of adding something inside the document modifies it. 

Manipulating the document the document should be a DBMS task, done right before storing or after retrieving it from disk.

Reply:
Hi @lamba92, I was looking into the test cases of yours and got the problem you are facing. You can take a look (if you have not already) at `EntityConverter` [here](https://nitrite.dizitart.com/java-sdk/repository/mapper/index.html#entityconverter).

May be, you don't need the kotlinx.serialization here at all. `EntityConverter` would be much simpler and efficient solution for you. It would be faster as there is no reflection involved.

Here is the code to use it, based on your [test criteria](https://github.com/lamba92/nitrite-java/blob/a911586788d89865153eabc74e75bf5a13f35397/potassium-nitrite/src/test/kotlin/org/dizitart/kno2/KotlinXSerializationMapperTest.kt#L239).

```kotlin
// test data class
data class DeepPutTestData(
    val name: String?,
    val userData: Map<String, String>?
) {
    companion object {
        fun create() = DeepPutTestData(
            name = "aString",
            userData = mapOf("the.key.to.split" to "aValue")
        )
    }
}

// converter class
class DeepPutTestConverter : EntityConverter<DeepPutTestData> {
    override fun getEntityType(): Class<DeepPutTestData> {
        return DeepPutTestData::class.java
    }

    override fun toDocument(entity: DeepPutTestData, nitriteMapper: NitriteMapper): Document {
        return Document.createDocument("name", entity.name)
            .put("userData", entity.userData)
    }

    @Suppress("UNCHECKED_CAST")
    override fun fromDocument(document: Document, nitriteMapper: NitriteMapper): DeepPutTestData {
        val deepPutTest = DeepPutTestData(
            name = document.get("name") as String?,
            userData = document.get("userData") as Map<String, String>?
        )
        return deepPutTest
    }
}

// test method
@Test
fun testDeepPut() {
    val db = nitrite {
        registerEntityConverter(DeepPutTestConverter()) // register your converter here
    }
    val repository = db.getRepository<DeepPutTestData>()
    val deepPutTestData = DeepPutTestData.create()
    repository.insert(deepPutTestData)
    val found = repository.find().firstOrNull()
    assertEquals(deepPutTestData, found)
}
```

The above code will not trigger `deepPut()` in the document inside your repository.

Reply:
Just following up. Did the `EntityConverter` solution solve your problem?